### ISSUE TITLE
AmigaOS4: crash on exit when free shader's resources and when close audio devices

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/Eldritch#7]
So there 2 crashes one coming after another when we about to shut down the game and exit from.

First one come from freeing shader resources: ShaderManagerFreeShaders->ShaderProgram->crash in ogles2.library

The second crash comes when we about to free resources of openal/alut, with output in console:

```AL lib: (EE) alc_cleanup: 1 device not closed.```

And have a warning window with "parent process has tried to exit before all children have". Error process "alsoft-mixer".

And stack tracepoint out on "ALCplaybackAHI_mixerProc"

**Comment by kas1e** [comment]
About crash in audio free stuff, interesting that if I fully comment out function OpenALAudioSystem::~OpenALAudioSystem(), it then says on exit that 2 devices not closed, and then the same crashes. While when all code in place, it says that 1 device not closed and the same crashes.

**Comment by ptitSeb** [comment]
There is only on `alcOpenDevice(...)` in the code: `Code/Libraries/Audio/src/openalaudiosystem.cpp` line 21.
Add a printf there, to check how many time it's called?

**Comment by kas1e** [comment]
Strange, putting print there I only see it 1 time. But commenting out the whole free function, I got from alc_cleanup: 2 devices not closed. Probably somehow related to openal i use.. will ask the author of port about.

Maybe now better check why this one https://github.com/ptitSeb/Eldritch/issues/6 happen?

**Comment by kas1e** [comment]
PS. Tried the other version of openal (some older port) and it didn't show that crash. So chances are high that it's a bug in our fresh openal port, let the ticket be open for the meantime anyway

**Comment by ptitSeb** [comment]
Yeah, I don't think Eldritch is the culprit here, not for OpenAL stuff.
